manager/allocator/networkallocator/networkallocator.go

```diff
@@ -376,593 +28,81 @@ func OnInit(options *ServiceAllocationOpts) {
+	IsAllocated(n *api.Network) bool
 
-		if s.Endpoint == nil || len(s.Endpoint.VirtualIPs) == 0 {
-			return true
-		}
+	// Allocate allocates all the necessary resources both general
+	// and driver-specific which may be specified in the NetworkSpec
+	Allocate(n *api.Network) error
 
-		// If the spec has networks which don't have a corresponding VIP,
-		// the service needs to be allocated.
-	networkLoop:
-		for _, net := range specNetworks {
-			for _, vip := range s.Endpoint.VirtualIPs {
-				if vip.NetworkID == net.Target {
-					continue networkLoop
-				}
-			}
-			return true
-		}
-	}
+	// Deallocate frees all the general and driver specific resources
+	// which were assigned to the passed network.
+	Deallocate(n *api.Network) error
 
-	// If the spec no longer has networks attached and has a vip allocated
-	// from previous spec the service needs to allocated.
-	if s.Endpoint != nil {
-	vipLoop:
-		for _, vip := range s.Endpoint.VirtualIPs {
-			if na.IsVIPOnIngressNetwork(vip) && IsIngressNetworkNeeded(s) {
-				continue vipLoop
-			}
-			for _, net := range specNetworks {
-				if vip.NetworkID == net.Target {
-					continue vipLoop
-				}
-			}
-			return true
-		}
-	}
+	//
+	// Service Allocation
+	//
 
-	// If the endpoint mode is DNSRR and allocator has the service
-	// in VIP allocated set then we return to be allocated to make
-	// sure the allocator triggers networkallocator to free up the
-	// resources if any.
-	if s.Spec.Endpoint != nil && s.Spec.Endpoint.Mode == api.ResolutionModeDNSRoundRobin {
-		if _, ok := na.services[s.ID]; ok {
-			return true
-		}
-	}
+	// ServiceNeedsAllocation returns true if the passed service
+	// needs to have network resources allocated/updated.
+	ServiceNeedsAllocation(s *api.Service, flags ...func(*ServiceAllocationOpts)) bool
 
-	if (s.Spec.Endpoint != nil && len(s.Spec.Endpoint.Ports) != 0) ||
-		(s.Endpoint != nil && len(s.Endpoint.Ports) != 0) {
-		return !na.portAllocator.isPortsAllocatedOnInit(s, options.OnInit)
-	}
-	return false
-}
+	// ServiceAllocate allocates all the network resources such as virtual
+	// IP and ports needed by the service.
+	ServiceAllocate(s *api.Service) (err error)
 
-// IsNodeAllocated returns if the passed node has its network resources allocated or not.
-func (na *NetworkAllocator) IsNodeAllocated(node *api.Node) bool {
-	// If the node is not found in the allocated set, then it is
-	// not allocated.
-	if _, ok := na.nodes[node.ID]; !ok {
-		return false
-	}
+	// ServiceDeallocate de-allocates all the network resources such as
+	// virtual IP and ports associated with the service.
+	ServiceDeallocate(s *api.Service) error
 
-	// If no attachment, not allocated.
-	if node.Attachment == nil {
-		return false
-	}
+	// HostPublishPortsNeedUpdate returns true if the passed service needs
+	// allocations for its published ports in host (non ingress) mode
+	HostPublishPortsNeedUpdate(s *api.Service) bool
```

Comment:
Note for the future: We should eventually refactor the code to remove this interface method. See  #2143.